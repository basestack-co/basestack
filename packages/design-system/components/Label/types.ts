export type LabelVariant =
  | "success"
  | "default"
  | "info"
  | "warning"
  | "danger"
  | "light";
export type LabelSize = "normal" | "medium" | "small";
