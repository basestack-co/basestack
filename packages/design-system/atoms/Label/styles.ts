import styled from "styled-components";
import { rem } from "polished";

export const StyledLabel = styled.div<{ variant: "success" | "default" }>`
  display: flex;
  padding: ${rem("4px")} ${rem("6px")};
  border-radius: 4px;
  font-size: ${rem("14px")};
  line-height: ${rem("16px")};
  font-weight: 400;
  color: ${({ theme, variant }) =>
    variant === "success" ? theme.colors.white : theme.colors.gray600};
  background-color: ${({ theme, variant }) =>
    variant === "success" ? theme.colors.green400 : theme.colors.gray200};
`;
