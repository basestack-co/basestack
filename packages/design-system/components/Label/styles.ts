import styled, { css } from "styled-components";
import { space, flexbox, layout, compose } from "styled-system";
import { rem } from "polished";
import { LabelVariant, LabelSize } from "./types";

const getGap = (size: LabelSize) => {
  const styles = {
    small: {
      gap: `${rem("6px")}`,
    },
    normal: {
      gap: `${rem("8px")}`,
    },
    medium: {
      gap: `${rem("10px")}`,
    },
  };

  return styles[size];
};

const getSizeStyles = (size: LabelSize) => {
  const styles = {
    small: {
      padding: `${rem("4px")} ${rem("6px")}`,
      fontSize: `${rem("12px")}`,
      lineHeight: `${rem("16px")}`,
      fontWeight: 400,
    },
    normal: {
      padding: `${rem("6px")} ${rem("8px")}`,
      fontSize: `${rem("14px")}`,
      lineHeight: `${rem("16px")}`,
      fontWeight: 400,
    },
    medium: {
      padding: `${rem("8px")} ${rem("10px")}`,
      fontSize: `${rem("16px")}`,
      lineHeight: `${rem("18px")}`,
      fontWeight: 400,
    },
  };

  return styles[size];
};

export const StyledLabel = styled.div<{
  variant: LabelVariant;
  size: LabelSize;
  isTranslucent: boolean;
  hasChildren: boolean;
  isUppercase: boolean;
}>`
  display: flex;
  border-radius: 4px;
  ${({ size }) => getSizeStyles(size)};
  ${({ theme, variant, isTranslucent }) =>
    isTranslucent
      ? theme.label.translucent[variant]
      : theme.label.solid[variant]};
  ${({ hasChildren, size }) =>
    hasChildren &&
    css`
      align-items: center;
      ${getGap(size)};
    `}
  ${({ isUppercase }) =>
    isUppercase &&
    css`
      text-transform: uppercase;
    `}
  ${compose(flexbox, space, layout)};
`;

export const LabelDotContainer = styled.div<{
  variant: LabelVariant;
  size: LabelSize;
  isUppercase: boolean;
}>`
  display: flex;
  align-items: center;
  font-size: ${({ size }) => getSizeStyles(size).fontSize};
  font-weight: ${({ size }) => getSizeStyles(size).fontWeight};
  line-height: ${({ size }) => getSizeStyles(size).lineHeight};
  ${({ isUppercase }) =>
    isUppercase &&
    css`
      text-transform: uppercase;
    `}
  ${compose(flexbox, space, layout)};
`;

export const LabelDot = styled.div<{
  variant: LabelVariant;
  isTranslucent: boolean;
}>`
  display: inline-flex;
  flex-shrink: 0;
  width: ${rem("14px")};
  height: ${rem("14px")};
  margin-right: ${rem("4px")};
  border-radius: 50%;
  ${({ theme, variant, isTranslucent }) =>
    isTranslucent
      ? theme.label.translucent[variant]
      : theme.label.solid[variant]};
`;
