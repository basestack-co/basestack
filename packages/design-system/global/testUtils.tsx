import { type RenderOptions, render } from "@testing-library/react";
import type React from "react";
// Styles
import { ThemeProvider } from "styled-components";
import theme from "../theme/lightTheme";

interface Props {
  children: React.ReactNode;
}

const WithThemeProvider = ({ children }: Props) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const renderWithTheme = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: WithThemeProvider, ...options });

export { renderWithTheme };
