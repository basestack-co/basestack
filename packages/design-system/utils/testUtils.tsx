import defaultTheme from "@basestack/design-system/theme/lightTheme";
import isPropValid from "@emotion/is-prop-valid";
import { RenderResult, render } from "@testing-library/react";
import React from "react";
import { StyleSheetManager, ThemeProvider } from "styled-components";

const Provider = ({
  children,
  theme = defaultTheme,
}: {
  children: React.ReactNode;
  theme: any;
}) => (
  <ThemeProvider theme={theme}>
    <StyleSheetManager shouldForwardProp={isPropValid}>
      {children}
    </StyleSheetManager>
  </ThemeProvider>
);

export const renderWithTheme = (
  ui: React.ReactElement,
  options?: any,
): RenderResult => render(ui, { wrapper: Provider, ...options });
