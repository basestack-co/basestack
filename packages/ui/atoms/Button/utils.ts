import { Size, Target, Type, Variant } from "@/components/atoms/Button/types";
import { css, DefaultTheme } from "styled-components";
import { darken, rem, transparentize } from "polished";

export const handleButtonSize = (size?: Size) => {
  switch (size) {
    case "large":
      return css`
        height: ${rem("60px")};
        font-size: ${rem("18px")};
        line-height: ${rem("25px")};
        padding: 0 ${rem("40px")};
      `;
    case "medium":
      return css`
        height: ${rem("54px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("34px")};
      `;
    case "normal":
      return css`
        height: ${rem("48px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("30px")};
      `;
    default:
    case "small":
      return css`
        height: ${rem("46px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("24px")};
      `;
    case "xSmall":
      return css`
        height: ${rem("42px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("20px")};
      `;
  }
};

export const handleButtonsColors = (theme: DefaultTheme) => ({
  primary: {
    outline: {
      color: theme.colors.primary,
      borderColor: theme.colors.primary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.primary,
    },
  },
  secondary: {
    outline: {
      color: theme.colors.secondary,
      borderColor: theme.colors.secondary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.secondary,
    },
  },
  tertiary: {
    outline: {
      color: theme.colors.gray3,
      borderColor: theme.colors.gray3,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: theme.colors.gray6,
    },
  },
  neutral: {
    outline: {
      color: theme.colors.gray2,
      borderColor: theme.colors.gray2,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: "transparent",
    },
  },
  success: {
    outline: {
      color: theme.colors.success,
      borderColor: theme.colors.success,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.success,
    },
  },
  warning: {
    outline: {
      color: theme.colors.warning,
      borderColor: theme.colors.warning,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.warning,
    },
  },
  danger: {
    outline: {
      color: theme.colors.error,
      borderColor: theme.colors.error,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.error,
    },
  },
});

export const darkenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return darken(
    amount || 0.1,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};

export const lightenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return transparentize(
    amount || 0.6,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};
