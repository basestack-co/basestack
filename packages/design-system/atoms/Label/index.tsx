import React, { memo } from "react";
import { SpaceProps } from "styled-system";
import { StyledLabel } from "./styles";

interface LabelProps extends SpaceProps {
  /**
   * Label variants
   */
  variant?: "success" | "default";
  /**
   * Label text
   */
  text: string;
  /**
   * Test id
   */
  testId?: string;
}

const Label = ({
  variant = "default",
  text,
  testId = "label-container",
  ...props
}: LabelProps) => (
  <StyledLabel data-testid={testId} variant={variant} {...props}>
    {text}
  </StyledLabel>
);

export default memo(Label);
