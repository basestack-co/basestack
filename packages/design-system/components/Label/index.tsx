import { memo, ReactNode } from "react";
import { SpaceProps, FlexboxProps, LayoutProps } from "styled-system";
import { LabelDot, LabelDotContainer, StyledLabel } from "./styles";
import { LabelVariant, LabelSize } from "./types";

export interface LabelProps extends SpaceProps, FlexboxProps, LayoutProps {
  /**
   * Label variants
   */
  variant?: LabelVariant;
  /**
   * Label text
   */
  text: string;
  /**
   * Test id
   */
  testId?: string;
  /**
   * Label size
   */
  size?: LabelSize;
  /**
   * Changes from solid colors to transparent if true
   */
  isTranslucent?: boolean;
  /**
   * Changes the styles of the label
   */
  type?: "dot" | "default";
  /**
   * Adds react component for ex icon button to clear label
   */
  children?: ReactNode;
  /**
   * Changes text to uppercase
   */
  isUppercase?: boolean;
}

const Label = ({
  text,
  size = "normal",
  type = "default",
  variant = "default",
  isUppercase = false,
  isTranslucent = false,
  testId = "label-container",
  children,
  ...props
}: LabelProps) => {
  if (type === "dot") {
    return (
      <LabelDotContainer
        isUppercase={isUppercase}
        variant={variant}
        size={size}
        {...props}
      >
        <LabelDot variant={variant} isTranslucent={isTranslucent} />
        <span>{text}</span>
      </LabelDotContainer>
    );
  }

  return (
    <StyledLabel
      data-testid={testId}
      variant={variant}
      size={size}
      isTranslucent={isTranslucent}
      isUppercase={isUppercase}
      hasChildren={!!children}
      {...props}
    >
      <span>{text}</span>
      {children}
    </StyledLabel>
  );
};

export default memo(Label);
