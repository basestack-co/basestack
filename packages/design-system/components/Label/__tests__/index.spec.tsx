import { cleanup } from "@testing-library/react";
import { renderWithTheme } from "../../../utils/testUtils";
import Label from "..";

describe("Label tests", () => {
  afterEach(cleanup);

  test("should render Label correctly", () => {
    const { asFragment } = renderWithTheme(
      <Label text="label" variant="default" />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render isTranslucent Label", () => {
    const { asFragment } = renderWithTheme(
      <Label text="label" variant="info" isTranslucent />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
