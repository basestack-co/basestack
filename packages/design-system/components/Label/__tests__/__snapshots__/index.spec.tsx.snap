// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label tests render Label correctly 1`] = `
<DocumentFragment>
  .c0 {
  display: flex;
  border-radius: 4px;
  padding: 0.25rem 0.375rem;
  font-size: 0.875rem;
  line-height: 1rem;
  font-weight: 400;
  color: #545454;
  background-color: #E2E2E2;
}

<div
    class="c0"
    data-testid="label-container"
    variant="default"
  >
    label
  </div>
</DocumentFragment>
`;

exports[`Label tests render isTranslucent Label 1`] = `
<DocumentFragment>
  .c0 {
  display: flex;
  border-radius: 4px;
  padding: 0.25rem 0.375rem;
  font-size: 0.875rem;
  line-height: 1rem;
  font-weight: 400;
  color: #1E54B7;
  background-color: rgba(39,110,241,0.1);
}

<div
    class="c0"
    data-testid="label-container"
    variant="info"
  >
    label
  </div>
</DocumentFragment>
`;
